sphinx_docs/build_docs/html/_sources/parameters.rst.txt
.. _parameters:

**********
Parameters
**********

.. _parametersParametersWindows:

Parameters Windows
==================

In the parameters window (see :numref:`parameters-windows-screenshot` ), the user can edit the output files name for the eyeTracker, audio, video and report files.

All parameters are saved in config.ini (see :ref:`simulationConfig` section).

**Important : All modification of file name will be applied only for the next recording.**

.. _parameters-windows-screenshot:
.. figure:: ./images/parameters_windows.png
   :alt: Parameters Window
   :width: 70.0%

   Parameters Window

.. _parametersConfigFile:

Configuration file
==================

All parameters in the parameters window are saved in the config.ini file.


Initial setup
-------------

The Initial setup is when the user run Eye Got for the first time. A popup inform the user that the initial setup is done, then, it can restart Eye Got It.

For more information about Eye Got It, see :ref:`compilationInstaller` in :ref:`compilation` section.


Action Units
============

Please see :ref:`actionUnits` section.

.. _parametersHardwareTest:

Hardware Test
=============

.. _parametersAudioTest:

Audio Test
----------

In the Sound test window (see :numref:`audio-test-screenshot`), the user can select the input for the
audio and test it. The user can also change the audio parameters.

The audio is not saved into an audio file.

.. _audio-test-screenshot:
.. figure:: ./images/audio_test.png
   :alt: Audio Test Windows
   :width: 70.0%

   Audio Test Windows

.. _parametersVideoTest:

Video Test
----------

When the user select the "Video Test" button, a new window is displayed
and show the webcam output. The user can stop the display by closing the
window or by pressing "q" or "enter" on the keyboard. This window only show the
webcam output and doesn’t record anything.


.. _parametersEyeTrackerTest:

Eye Tracker Test
----------------

In the Eye Tracker Test, you cant check if the Eye Tracker is correctly calibrated (See :numref:`eyeTracker-test-screenshot`). 

When you click on the "start" button, tle lines are hiding before to be showing, the test is started. You can stop the test by click on the "stop" button or press the "ENTER". 
The Eye tracker Gazes are displayed on the windows. For more information about the Eye Tracker test, please see the :ref:`eyeTracker` section.

When you close the windows the temporary Eye Tracker data are deleted.

.. _eyeTracker-test-screenshot:
.. figure:: ./images/eyeTrackerTestCalibration.png
   :alt: Eye Tracker Test Calibration
   :width: 100.0%

   Eye Tracker Test Calibration


Changelog.txt
Changelog :

6.0: 2021-12-15:
Validation of Eye Tracker recording
Calculating the Eye-voice Span
Combining EVS and Validation
updating the documentation accordingly 
Adding installation script 

5.1: 2021-07-01:
Final Eye Got It version
Bug fixed
Documentation improved

5.0: 2021-06-30
Final Eye Got It version
Bug fixed
Improve User Interaction
Add final Documentation

4.4: 2021-06-25
add sphinx documentation in Eye Got it
bug fixed in report
improve user interaction
improve initial setup
add action units during report generation
manage action units

4.3: 2021-06-23
bug correction
report generation improved : multiple report with multiple algorithm (buscher and nystrom) and with multiple machine learning support
simulation config, add audio and video test
in report folder one folder by algorithm and one folder by machine learning inside
update report without delete report

4.2.1: 2021-06-21 
Bug fixed in eye Tracker calibration test 

4.2: 2021-06-18
Add Eye Tracker test Calibration
User interaction improved
support record session and multiple report generation
bug correction
Add Nystrom algorithm

4.1: 2021-06-14
Bug fixed for audio and opencv
Disable Eye Tracker test because it's not necessary to has it
Bug fixed for text adaptation function and save word coordinate

4.0: 2021-06-11
Add machine learning
New report and simulation config windows
Opencv bug fixed
Add Tobii SDK pro for Eye Tracker Pro
Bug fixed


3.3: 2021-05-26
Change CSV format 
Bug fixed 
Add Documentation 
Create Machine learning code (not integrated)


3.2: 2021-04-26
Changing MCQ CSV File : now in columns + header
Changing Answer CSV MCQ File : now in columns + header
Changing User CSV File : now in columns + header
Bug correction in user management


3.1: 2021-04-26
Bug in choose text to create MCQ fixed
Bug in MCQ creator answer fixed
New report.txt
MCQ viewer in simulation improved and reworked (new display and current number question display)
Display current number question in MCQ creator
Database bug fixed
MCQ Creation reworked
Added OpenFace for to process video
Video processed with OpenFace + video popup
Report Function added new features
Ask "Open the Folder at the end" by checkbox when generate report


3.0-1: 2021-04-19
Bug in MCQ save answer fixed
Bug in Parameter windows fixed


3.0: 2021-04-19
MCQ Management improved and reworked
Parameters windows improved and reworked
New Database type supported
Random simulation
report improved and reworked
calcul speed in degrees of visual angle /s
Help windows added
EyeTracker Head capturing (only for eyetracker 5)
Eye Tracker 5 supported
Video processing to get faces and eyes position (and display estimated time)
Nystrom algorithm implemented but not used, we used Buscher instead
Better interface navigation


2.0: 2021-04-02
MCQ Management (create and edit)
Database Management (create and edit)
Video record by webcam added
Sound bug fixed
Buscher algorithm implementation
gaze and fixation save to csv
Create Report added
Simulation bug fixed
User Management bug fixed
Parameters add "video test" function
Parameters sound "test bug" fixed
eye tracker support (capture data)
eye tracker save page to csv
overall windows navigation improved
save UNIX Time for page and sound and video


1.2 : 2021-03-22
database and user folder now are in User personnal folder 
sound test (record and listen) added in parameter windows
mcq answer added in mcq config


1.1 : 2021-03-19
New Simulation windows
display label "recording" when simulation start
display "How to use simulation mode"
Database management
page gestion for text if the text is too long
screenshot of each text page in simulation mode
bug correction
remove unnecessary features 
rework of save word coordonates feature
text display space line
clean code


1.0 : 2021-03-16
First Release 
save word position
adding mcq
user gestion
highlighting of words not understood 
sound recording added
select text in list and display it
switch from tkinter to pyQt5 
save mcq answers
parameters gestion (config.ini and parameters windows)




eye_got_it/eye_got_it/Config.py
#Config.py => eye got it param contain global variables used, config import and theme

#import python Module
import time, sys, os, io
from PyQt5 import QtCore, QtWidgets, QtGui
#import others Modules
import ConfigFileManagement, Database, Function, Welcome

DEBUG=False # debug session

def currentTime(): # return the current time
    return str(time.strftime('%H_' '%M_' '%S'))

def currentDate(): # return the current date
    return str(time.strftime('%Y_' '%m_' '%d_'))

#config import
CONFIG_IMPORTED=False
CONFIG_IMPORTED_ERROR=False
configParser= None
initial_setup="" # first run of eye got it

#eye got it folder
EyeGotItFolder=""# main data folder
DatabaseFolder=""
UsersFolder=""
ReportFolder=""
PATH_IMAGE = os.path.join("Pictures")

#eye got it
VERSION=""
LOGO_IMAGE=os.path.join(PATH_IMAGE , "logo.png")

#text choose by user location path
TXT_IMPORT=None

#eyeTracker
eyeTrackerSdkPro = False #eyeTracker gaming (false) or Pro (true)
eyeTrackerList = ["Tobii Pro Nano", "Tobii Gaming"]
interactionLibrary = "eyeTracker"
sdkPro = "eyeTrackerPro"
onlyBuscher = False
eyeTrackerAlgorithm = ['buscher','nystrom']

#test calibration eyeTracker
eyeTrackerCalibration="tempGazes.csv"
eyeTrackerHeadCalibration="tempHeadGazes.csv"

#screen variable
SCREEN_WIDTH=None
SCREEN_HEIGHT=None
SCREEN_WIDTH=None
PHYSICAL_SCREEN_HEIGHT=None
PHYSICAL_SCREEN_WIDTH=None
SCREEN_WIDTH_SIMULATION=1366# DO NOT MODIFY
SCREEN_HEIGHT_SIMULATION=650# DO NOT MODIFY

#video Input
videoInput=[]
actionUnitMinIntensity=2 #minimal to maximal => 1 to 5

#audio Input
audioInput = []

#simulation folder name
SAVE_DATA_FOLDER = ""
SAVE_DATA_MAIN_FOLDER = ""
TIME_FOLDER=currentTime()#init time session
NAME_SAVE_FOLDER = currentDate() + TIME_FOLDER#init saving folder

#Log file
LOG_FILE="eye_got_it.log"
LOG_TYPE="DEBUG"

#theme and police
THEME=""
colorFont = ""
colorText = ""
colorSelect = ""
background=""
#police="Noto Sans"
police=""
fontSize=20
fontSizeText=25
fontType="normal"

#audio parameters
rate=""
channels=""
chunk=""
sound_out=""

#databases
databaseName=""
database=""

#MCQ Answers by column :
MCQAnswerByColumn = 4

#ActionUnits by column :
ActionUnitsByColumn = 4

#parameters
screenshotName=""
screenshotExtension="jpg"
eyeTrackerHeadData=""
eyeTrackerData=""
eyeTrackerDataFile="csv"
videoOut=""
videoUNIXStart=""
audioUNIXStart=""
HelpIcon = ""

#Report files
eyeTrackerGlissades = ""#"glissades"
eyeTrackerSaccades = ""#"saccades"
eyeTrackerFixations = ""#"fixations"
eyeTrackerVelocityGazes = ""#"velocityGazes"
eyeTrackerScreenshot = ""#"EyeTracker"
wordScreenshot = ""#"Word_Page"

#list of database
DatabaseList=[]

#list of users
UsersList=[]

#hardware test
audioTest = True
videoTest = True
eyeTrackerTest = True

#HTML documentation
documentationLocation = os.path.join(os.getcwd(),"docs","html")
documentationHTML=False

#OpenFace => DO NOT EDIT Unless a new version supports other action units 
openFaceActionUnit=[1, 2, 4, 5, 6, 7, 9, 10, 12, 14, 15, 17, 20, 23, 25, 26, 28, 45]

#Eye Got It Documentation
if os.path.isdir(documentationLocation):
    documentationHTML= True
    documentationWelcome = os.path.join(documentationLocation,"welcome.html")
    documentationSimulation = os.path.join(documentationLocation,"simulation.html")
    documentationParameters = os.path.join(documentationLocation,"parameters.html")
    documentationParametersAudioTest = os.path.join(documentationLocation,"parameters.html#audio-test")
    documentationUsers = os.path.join(documentationLocation,"user.html")
    documentationDatabase = os.path.join(documentationLocation,"database.html")
    documentationReport = os.path.join(documentationLocation,"report.html")
    documentationActionUnits = os.path.join(documentationLocation,"actionUnits.html")
    documentationMCQ = os.path.join(documentationLocation,"mcq.html")

#import the config
def ImportConfig():
    global eyeTrackerHeadData,videoOut, VERSION, configParser, screenshotName, screenshotExtension
    global textPart, mcqAnswer, textPosition, THEME, CONFIG_IMPORTED, CONFIG_IMPORTED_ERROR
    global rate, channels, chunk, soundOut,initial_setup, eyeTrackerData, eyeTrackerGlissades, eyeTrackerSaccades
    global eyeTrackerFixations, eyeTrackerVelocityGazes, eyeTrackerScreenshot, wordScreenshot

    try:
        configParser= ConfigFileManagement.ConfigFileRead('config.ini')
        THEME=str(configParser.GetParameters("THEME", 'theme'))
        
        screenshotName = str(configParser.GetParameters('SYSTEM', 'screenshot_name'))
        VERSION = str(configParser.GetParameters('SYSTEM', 'version'))
        theme(THEME)
        rate = int(configParser.GetParameters('SOUND', 'rate'))
        channels = int(configParser.GetParameters('SOUND', 'channels'))
        chunk = int(configParser.GetParameters('SOUND', 'chunk'))
        soundOut = str(configParser.GetParameters('SOUND', 'sound_out'))
        videoOut = str(configParser.GetParameters('VIDEO', 'video_out'))
        mcqAnswer=str(configParser.GetParameters('USER', 'mcq_answer'))
        textPosition=str(configParser.GetParameters('USER', 'text_position'))
        textPart=str(configParser.GetParameters('USER', 'text_part'))
        eyeTrackerHeadData = str(configParser.GetParameters('EYETRACKER', 'eye_tracker_head_data'))
        eyeTrackerData = str(configParser.GetParameters('EYETRACKER', 'eye_tracker_data'))
        initial_setup=(configParser.GetParameters('OTHER', 'initial_setup'))

        eyeTrackerGlissades = str(configParser.GetParameters('REPORT', 'eyeTracker_Glissades'))
        eyeTrackerSaccades = str(configParser.GetParameters('REPORT', 'eyeTracker_Saccades'))
        eyeTrackerFixations = str(configParser.GetParameters('REPORT', 'eyeTracker_Fixations'))
        eyeTrackerVelocityGazes = str(configParser.GetParameters('REPORT', 'eyeTracker_VelocityGazes'))
        eyeTrackerScreenshot = str(configParser.GetParameters('REPORT', 'eyeTracker_Screenshot'))
        wordScreenshot = str(configParser.GetParameters('REPORT', 'word_Screenshot'))

        databaseListCreate()
        userListCreate()
        CONFIG_IMPORTED=True

    except:
        CONFIG_IMPORTED=False
        CONFIG_IMPORTED_ERROR=True
    
    

#import database
def CreateDatabase(Name):
    global database,databaseName
    databaseName=Name
    if os.path.isdir(os.path.join(DatabaseFolder,Name)) :
        database=Database.Database(os.path.join(DatabaseFolder,Name))
        database.CreateConfig()
        return True
    else :
        return False


#import theme
def theme(Theme):
    global colorFont, colorText, colorSelect,background, HelpIcon

    if Theme == "Normal":
        colorFont = "white"
        colorText = "black"
        colorSelect = "black"
        background = "white"
        HelpIcon = os.path.join("Pictures","help.png")
        return True

    if Theme == "Dark":
        colorFont = "black"
        colorText = "white"
        colorSelect = "white"
        background = "black"
        HelpIcon = os.path.join("Pictures","help_dark.png")
        return True
    else :
        return False


def fileExist(fileToTest): # detect if file in folder exist
    try:
        with open(fileToTest,'r',encoding='latin1') as filename:            
            return  True #file exist

    except: #no file
        Log.InfoSaveLog("warning",str(fileToTest + "not exist"))       
        return False # return error file no found


#import list of databases
def databaseListCreate():
    global DatabaseList

    listDatabase = os.listdir(DatabaseFolder)
    DatabaseList=[]
    for elem in listDatabase:
        if len(elem.split('.'))==1 and os.path.isfile(os.path.join(DatabaseFolder,elem,"config.ini")):
            DatabaseList.append(elem)
    DatabaseList.sort()


#import list of users
def userListCreate():
    global UsersList

    listUser = os.listdir(UsersFolder)
    UsersList=[]
    for elem in listUser:
        if len(elem.split('.'))==1 and os.path.isfile(os.path.join(UsersFolder,elem,str(elem + ".csv"))):
            UsersList.append(elem)
    UsersList.sort()